core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/keymanager/imports/ImportKeystoresHandler.java

```diff
@@ -0,0 +1,243 @@
+        context.fail(SERVER_ERROR, e);
+      }
+      return;
+    }
+
+    // extract pubkeys to import first
+    final List<String> pubkeysToImport = new ArrayList<>();
+    try {
+      pubkeysToImport.addAll(
+          parsedBody.getKeystores().stream()
+              .map(json -> new JsonObject(json).getString("pubkey"))
+              .map(
+                  key ->
+                      key.startsWith("0x")
+                          ? key
+                          : "0x" + key) // always use 0x prefix for comparisons
+              .collect(Collectors.toList()));
+    } catch (Exception e) {
+      context.fail(BAD_REQUEST, e);
+      return;
+    }
+
+    // read slashing protection data if present and import data matching keys to import only
+    if (slashingProtection.isPresent()) {
+      try {
+        final InputStream slashingProtectionData =
+            new ByteArrayInputStream(
+                parsedBody.getSlashingProtection().getBytes(StandardCharsets.UTF_8));
+        slashingProtection.get().importDataWithFilter(slashingProtectionData, pubkeysToImport);
+      } catch (Exception e) {
+        context.fail(BAD_REQUEST, e);
+        return;
+      }
+    }
+
+    final Set<String> existingPubkeys = artifactSignerProvider.availableIdentifiers();
+    final List<ImportKeystoreResult> results = new ArrayList<>();
+    for (int i = 0; i < parsedBody.getKeystores().size(); i++) {
+      final String pubkey = pubkeysToImport.get(i);
+      try {
+        final String jsonKeystoreData = parsedBody.getKeystores().get(i);
+        final String password = parsedBody.getPasswords().get(i);
+
+        if (existingPubkeys.contains(pubkey)) {
+          // keystore already loaded
+          results.add(new ImportKeystoreResult(ImportKeystoreStatus.DUPLICATE, null));
+        } else {
+          // new keystore to import
+          // 1. validate and decrypt the keystore
+          final BlsArtifactSigner signer = validateKeystore(jsonKeystoreData, password);
+          // 2. write keystore file to disk
+          createKeyStoreYamlFileAt(pubkey, jsonKeystoreData, password);
+          // 3. add the new signer to the provider to make it available for signing
+          artifactSignerProvider.addSigner(pubkey, signer).get();
+          // 4. finally, add result to API response
+          results.add(new ImportKeystoreResult(ImportKeystoreStatus.IMPORTED, null));
+        }
+      } catch (Exception e) {
+        cleanupImportedKeystoreFiles(List.of(pubkey));
+        results.add(
+            new ImportKeystoreResult(
+                ImportKeystoreStatus.ERROR, "Error importing keystore: " + e.getMessage()));
+      }
+    }
+
+    try {
+      context
+          .response()
+          .putHeader(CONTENT_TYPE, JSON_UTF_8)
+          .setStatusCode(SUCCESS)
+          .end(objectMapper.writeValueAsString(new ImportKeystoresResponse(results)));
+    } catch (Exception e) {
+      cleanupImportedKeystoreFiles(pubkeysToImport);
```

Comment:
If we error at this point, do we also need to clean up any keys from memory?
It looks like we could successfully execute `artifactSignerProvider.addSigner(pubkey, signer).get();` but then return a 500 response.

Reply:
good catch. I'll add a synchronous `removeSigner` method here. I will need it for the delete endpoint anyways